common/functions/src/aggregates/aggregate_retention.rs

```diff
@@ -0,0 +1,214 @@
+    }
+
+    fn accumulate(
+        &self,
+        place: StateAddr,
+        columns: &[common_datavalues::ColumnRef],
+        _validity: Option<&common_arrow::arrow::bitmap::Bitmap>,
+        input_rows: usize,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        let new_columns: Vec<&BooleanColumn> = columns
+            .iter()
+            .map(|column| Series::check_get(column).unwrap())
+            .collect();
+        for i in 0..input_rows {
+            for j in 0..self.events_size {
+                if new_columns[j as usize].get_data(i) {
+                    state.add(j);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn accumulate_row(
+        &self,
+        place: StateAddr,
+        columns: &[common_datavalues::ColumnRef],
+        row: usize,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        let new_columns: Vec<&BooleanColumn> = columns
+            .iter()
+            .map(|column| Series::check_get(column).unwrap())
+            .collect();
+        for j in 0..self.events_size {
+            if new_columns[j as usize].get_data(row) {
+                state.add(j);
+            }
+        }
+        Ok(())
+    }
+
+    fn serialize(&self, place: StateAddr, writer: &mut BytesMut) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        state.serialize(writer)
+    }
+
+    fn deserialize(&self, place: StateAddr, reader: &mut &[u8]) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        state.deserialize(reader)
+    }
+
+    fn merge(&self, place: StateAddr, rhs: StateAddr) -> Result<()> {
+        let rhs = rhs.get::<AggregateRetentionState>();
+        let state = place.get::<AggregateRetentionState>();
+        state.merge(rhs);
+        Ok(())
+    }
+
+    #[allow(unused_mut)]
+    fn merge_result(
+        &self,
+        place: StateAddr,
+        array: &mut dyn common_datavalues::MutableColumn,
+    ) -> Result<()> {
+        let state = place.get::<AggregateRetentionState>();
+        let builder: &mut MutableObjectColumn<JsonValue> = Series::check_get_mutable_column(array)?;
+        let mut vec: Vec<JsonValue> = vec![json!(0); self.events_size as usize];
+        if state.events & 1 == 1 {
+            vec[0] = json!(1);
+            for i in 1..self.events_size {
+                if state.events & (1 << i) != 0 {
+                    vec[i as usize] = json!(1);
+                }
+            }
+        }
+        builder.append_value(JsonValue::Array(vec));
+        Ok(())
+    }
+
+    fn accumulate_keys(
+        &self,
+        places: &[StateAddr],
+        offset: usize,
+        columns: &[common_datavalues::ColumnRef],
+        _input_rows: usize,
+    ) -> Result<()> {
+        for (row, place) in places.iter().enumerate() {
+            self.accumulate_row(place.next(offset), columns, row)?;
```

Comment:
Better to expand this function.

Inside the `accumulate_row` we need to `check_get` and build the columns, this may cost too much inside the loop.

Reply:
ok! nice advice